app/controllers/orderitems_controller.rb

```diff
@@ -0,0 +1,171 @@
+# frozen_string_literal: true
+
+class OrderitemsController < ApplicationController
+  before_action :require_login, only: [:mark_shipped]
+  before_action :find_oi, except: [:create]
+
+  def create
+    @cart = Order.find_by(id: session[:order_id])
+    product = Product.find_by(id: params[:product_id])
+    quantity = params[:orderitem][:quantity].to_i
+
+    # commenting this out - because of ensure_cart in application controller we don't need to check this
+    # if @cart.nil?
+    #   @cart = Order.new
+    #   @cart.status = "pending"
+    #   @cart.save
+    #   session[:order_id] = @cart.id
+    # end
+
+    if product.nil?
+      flash[:warning] = 'You must select a valid product'
+      redirect_back(fallback_location: root_path)
+      return
+    end
+
+    if product.merchant.id == session[:user_id]
+      flash[:warning] = 'You cannot add your own product to your cart'
+      redirect_back(fallback_location: root_path)
+      return
+    end
+
+    if quantity.nil? || quantity < 1
+      flash[:warning] = 'Cannot add 0 items'
+      redirect_back(fallback_location: root_path)
+      return
+    elsif product.stock.zero?
+      flash[:warning] = 'Item out of stock, cannot add to cart'
+      redirect_back(fallback_location: root_path)
+      return
+    elsif quantity > product.stock
+      flash[:warning] = "Only #{product.stock} items left in stock"
+      redirect_back(fallback_location: root_path)
+      return
+    end
+
+    if @orderitem = @cart.orderitems.find_by(product_id: product.id)
+      new_quant = quantity + @orderitem.quantity.to_i
+      @orderitem.update(quantity: new_quant)
+      flash[:success] = "Successfully updated #{@orderitem.product.name} quantity!"
+      redirect_to cart_path
+      return
+    end
+
+    @orderitem = Orderitem.new
+    @orderitem.order_id = @cart.id
+    @orderitem.product_id = product.id
+    @orderitem.quantity = quantity
+    @orderitem.shipped = false
+
+
+    if @orderitem.save!
+      flash[:success] = 'Product successfully added to cart!'
+      redirect_to cart_path
+      return
+    else
+      flash[:warning] = 'A problem occurred: could not add item to cart'
+      redirect_back(fallback_location: root_path)
+      return
+    end
+  end
```

Comment:
This will create a ton of bloat data! why not just store info in session until you need to save all of it? (as in during a checkout?)